### PR TITLE
Cypress UI tests

### PR DESCRIPTION
**Pull request by andresbott**
Migrate UI test to cypress

## Description

Remove the default selenium based UI tests, and replace them with cypress

The new tests 
a) follow the author flow verifying:
* page title is editable
* we can add a component on a page and publish that

b) verify the public page
* check the that the front page has a hero image
* check that search contains results

## Related Issue

jira.corp.adobe.com/browse/CQ-4354669

## Recordings of the verification

https://github.com/adobe/aem-guides-wknd/assets/1543913/646e96ea-be75-47dd-a692-5492c45311b1

https://github.com/adobe/aem-guides-wknd/assets/1543913/b58e1eeb-88e6-4988-9fb8-00f8111a7408

### CODE REVIEW

File: ui.tests/README.md [hunk 1]
```diff
@@ -1,106 +1,82 @@
+# Cypress sample module
 
-UI Testing module for your AEM application
-===
+Sample structure  for [Cypress](https://www.cypress.io) UI test module which conforms to
+AEM  Cloud Manager quality gate UI test conventions.
```
Comment (dulvac) [inline ui.tests/README.md:4]:
```suggestion
AEM Cloud Manager quality gate UI test conventions.
```

File: ui.tests/README.md [hunk 2]
```diff
@@ -1,106 +1,82 @@
...
-* `assembly-ui-test-docker-context.xml` Packages test project for AEMaaCS
-
-
-## Requirements
-
-* Maven
-* Latest version of Chrome and/or Firefox browser installed locally in default location
-* An AEM author instance running at http://localhost:4502
-* Sample application deployed on your AEM author instance (see [../README.md](../README.md))
+AEM provides an integrated suite of Cloud Manager quality gates to ensure smooth updates to custom applications,
```
Comment (dulvac) [inline ui.tests/README.md:8]:
```suggestion
AEM provides an integrated suite of Cloud Manager quality gates to ensure smooth updates to custom applications;
```

File: ui.tests/README.md [hunk 3]
```diff
@@ -1,106 +1,82 @@
...
+### Dockerfile
 
+Sample dockerfile is based on the `cypress/included` [image](https://hub.docker.com/r/cypress/included), which provides all the dependencies and the binaries
+to run cypress tests.
 
-## Docker execution
+### xvfb setup
 
-This project also provides Maven profiles to build and execute the tests using Docker
+>When running several Cypress instances in parallel, the spawning of multiple X11 servers at once can cause problems for some of them. In this case, you can separately start a single X11 server and pass the server's address to each Cypress instance using DISPLAY variable.
```
Comment (dulvac) [inline ui.tests/README.md:32]:
```suggestion
>When running several Cypress instances in parallel, the spawning of multiple X11 servers at once can cause problems for some of them. In this case, you can separately start a single X11 server and pass the server's address to each Cypress instance using the DISPLAY environment variable.
```

File: ui.tests/README.md [hunk 4]
```diff
@@ -1,106 +1,82 @@
...
 
-### Requirements
+The setup described in [the documentation](https://docs.cypress.io/guides/continuous-integration/introduction#In-Docker) 
+is implemented in `run.sh` as is used as entrypoint to the container.
```
Comment (dulvac) [inline ui.tests/README.md:35]:
```suggestion
is implemented in `run.sh`,  used as the _entrypoint_ to the container.
```

File: ui.tests/README.md [hunk 5]
```diff
@@ -1,106 +1,82 @@
...
-### Build test image
+### Locally (standalone)
 
-```
-mvn clean install -Pui-tests-docker-build
-```
+Refer to [test-module/README.md](test-module/README.md).
 
-will build Docker image `com.adobe.aem.guides-aem-guides-wknd.ui.tests/ui.tests` locally
+### Build and run test image
```
Comment (dulvac) [inline ui.tests/README.md:43]:
```suggestion
### Build and run the test image
```

**Review by dulvac — COMMENTED** [review]
Cool.
Minor comments

File: ui.tests/Dockerfile [hunk 6]
```diff
@@ -1,64 +1,31 @@
...
 # distributed under the License is distributed on an "AS IS" BASIS,
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-#
-# DO NOT MODIFY
-#
-FROM node:lts-slim
+FROM cypress/included:12.17.0
```
Comment (bisswanger) [inline ui.tests/Dockerfile:15]: why not 12.17.4 or even cypress 13? ;-)

File: ui.tests/Dockerfile [hunk 7]
```diff
@@ -1,64 +1,31 @@
...
     && apt -qqy --no-install-recommends install \
     python \
-    build-essential \
-    # for wait_for_grid script
-    && apt -qqy --no-install-recommends install \
-    curl \
-    jq
-# <<< End CloudManager Convention
+    build-essential
+# <<< End EAAS Convention \
```
Comment (bisswanger) [inline ui.tests/Dockerfile:23]: You did remove the Begin Convention part above >>>
Either keep or remove both

File: ui.tests/pom.xml [hunk 8]
```diff
@@ -18,29 +18,19 @@
...
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
     <modelVersion>4.0.0</modelVersion>
 
-    <parent>
-        <groupId>com.adobe.aem.guides</groupId>
-        <artifactId>aem-guides-wknd</artifactId>
-        <version>3.2.1-SNAPSHOT</version>
-        <relativePath>../pom.xml</relativePath>
-    </parent>
+    <groupId>com.adobe.cq.cloud</groupId>
```
Comment (bisswanger) [inline ui.tests/pom.xml:25]: hm, wouldn't we want to keep the parent relation etc?
As that's how it also looks like when generated using the archetype?
And is nicer if they create new versions with mvn release etc

File: ui.tests/test-module/package.json [hunk 9]
```diff
@@ -1,53 +1,33 @@
...
-  },
-  "license": "SEE LICENSE IN LICENSE file",
-  "keywords": [
-    "aem",
-    "cloud",
-    "aemaacs",
-    "selenium",
-    "wdio"
-  ],
+  "description": "test cypress on eaas",
```
Comment (bisswanger) [inline ui.tests/test-module/package.json:4]: hm, rather rename the package name/description to include WKND in some way?

File: ui.tests/test-module/cypress/support/wknd.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+import { RandomPageTitle } from './aem' // contains some utility functions
+
+// WKNDCreateContentPage creates a random content page, the page details can be accessed with this.wkndPage
+// NOTE: Don't use an arrow function as they don't have their own bindings to this, arguments, or super,
+
+Cypress.Commands.add('WKNDCreateContentPage', function (dir,title) {
```
Comment (bisswanger) [inline ui.tests/test-module/cypress/support/wknd.js:22]:
```suggestion
Cypress.Commands.add('WKNDCreateContentPage', function (dir, title) {
```

File: ui.tests/test-module/cypress/fixtures/example.json [hunk 11]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (bisswanger) [inline ui.tests/test-module/cypress/fixtures/example.json:1]: I think this is not used anywhere and we can remove it?

File: ui.tests/test-module/cypress/support/aem.js [hunk 12]
```diff
@@ -0,0 +1,283 @@
...
+        }
+    })
+})
+
+
+
+Cypress.Commands.add('AEMPathExists', function (baseUrl, path) {
+    const url = new URL(path, baseUrl)
+
+    console.log('COMMAND CALLED - START')
```
Comment (bisswanger) [inline ui.tests/test-module/cypress/support/aem.js:188]: remove the logging?

File: ui.tests/test-module/cypress/e2e/authoring.cy.js [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+
+  it('should be able to see a page and change page title', function () {
+    cy.AEMNavigateSites(testPage.path)
+
+    cy.AEMNavigatePageProperties(testPage.path)
+    cy.get('input[name="./jcr:title"]').clear()
+    cy.get('input[name="./jcr:title"]').type(testPage.title + '-edited')
+    cy.get('button[type="submit"]#shell-propertiespage-doneactivator').click()
+    cy.visit('/')
+    // navigate back and asset the value is stored
```
Comment (bisswanger) [inline ui.tests/test-module/cypress/e2e/authoring.cy.js:51]:
```suggestion
    // navigate back and assess the value is stored
```

**Review by bisswanger — APPROVED** [review]

File: ui.tests/Dockerfile (hunk 7 above)
Comment (andresbott) [reply ui.tests/Dockerfile:23]: https://github.com/adobe/aem-test-samples/blob/aem-cloud/ui-cypress/Dockerfile#L23

File: ui.tests/pom.xml (hunk 8 above)
Comment (andresbott) [reply ui.tests/pom.xml:25]: https://github.com/adobe/aem-test-samples/blob/aem-cloud/ui-cypress/pom.xml#L25

File: ui.tests/test-module/package.json (hunk 9 above)
Comment (andresbott) [reply ui.tests/test-module/package.json:4]: https://github.com/adobe/aem-test-samples/blob/aem-cloud/ui-cypress/test-module/package.json#L4

File: ui.tests/test-module/cypress/fixtures/example.json (hunk 11 above)
Comment (andresbott) [reply ui.tests/test-module/cypress/fixtures/example.json:1]: :+1:

**Review by davidjgonzalez — APPROVED** [review]

**Comment by davidjgonzalez** [comment]
@andresbott should the PR be moved out of Draft yet?

**Comment by andresbott** [comment]
@davidjgonzalez sorry, i missed the notification;
I checked again this tests on a local environment and all is green; this can be merged

**Comment by SachinMali** [comment]
@andresbott Can you please resolve the remaining conversations, we can merge it soon then, thanks!

**Comment by andresbott** [comment]
@SachinMali done; the points in the conversations were fixed; i just closed the conversations

**Comment by andresbott** [comment]
@SachinMali  as discussed I verified again the Local execution of all the steps to execute the UI tests against an AEMCS instance.
as well I updated the Readme accordingly

**Comment by SachinMali** [comment]
@andresbott Thanks, appreciated calling out it's AEMCS specific.

**Pull request merged**